internal/cri/opts/spec_linux_opts.go

```diff
@@ -106,11 +107,65 @@ func withMounts(osi osinterface.OS, config *runtime.ContainerConfig, extra []*ru
 				dst = mount.GetContainerPath()
 				src = mount.GetHostPath()
 			)
+
+			var statError error
+			mu := osManager.InflightStatsMutex()
+			inflightStatMap := osManager.InflightStats()
+
+			mu.Lock()
+			inflight, isPending := inflightStatMap[src]
+			if isPending {
+				// Stat for this path is already in progress. We wait on the
+				// condition variable. Wait() atomically unlocks the mutex and
+				// suspends the goroutine. When awakened, it re-locks the mutex
+				// before returning.
+				for !inflight.Done {
+					inflight.Cond.Wait()
+				}
+				statError = inflight.Err
+				mu.Unlock()
+			} else {
+				// There are no active Stats being made to this path. Create a
+				// record for other goroutines to find and wait on.
+				newInflight := osManager.NewInflightStat()
+				newInflight.Cond = sync.NewCond(mu)
+				inflightStatMap[src] = newInflight
+				mu.Unlock()
+
+				statCtx, statcancel := context.WithTimeout(ctx, osManager.StatTimeout())
+				resultChan := make(chan osinterface.StatResult, 1)
+				go func() {
+					_, err := osi.Stat(src)
+					resultChan <- osinterface.StatResult{Err: err}
+					close(resultChan)
+				}()
+
+				select {
+				case res := <-resultChan: // Stat completed
+					statError = res.Err
+				case <-statCtx.Done(): // timeout / mountpath could be stuck.
+					statError = fmt.Errorf("mountpath (%q) could be stuck", src)
+				}
+				statcancel()
+
+				mu.Lock()
+				newInflight.Err = statError
+				newInflight.Done = true
+
+				// Delete the src entry from the map so that subsequent requests
+				// can still call a new Stat to the src in case the underlying FS
+				// has recovered and can serve files.
+				delete(inflightStatMap, src)
+
+				newInflight.Cond.Broadcast()
+				mu.Unlock()
+			}
```

Comment:
It seems to me like we're leaking implementation details of `osManager` here. I would expect this method would call out to `osManager.Stat()`, and then the `osManager` will internally figure out whether to issue a new `stat` or bundle this request with one already in-flight.